src/vs/editor/browser/gpu/raster/glyphRasterizer.ts

```diff
@@ -0,0 +1,225 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Disposable } from '../../../../base/common/lifecycle.js';
+import { StringBuilder } from '../../../common/core/stringBuilder.js';
+import { FontStyle, TokenMetadata } from '../../../common/encodedTokenAttributes.js';
+import { ensureNonNullable } from '../gpuUtils.js';
+import type { IBoundingBox, IGlyphRasterizer, IRasterizedGlyph } from './raster.js';
+
+let nextId = 0;
+
+export class GlyphRasterizer extends Disposable implements IGlyphRasterizer {
+	public readonly id = nextId++;
+
+	private _canvas: OffscreenCanvas;
+	private _ctx: OffscreenCanvasRenderingContext2D;
+
+	private _workGlyph: IRasterizedGlyph = {
+		source: null!,
+		boundingBox: {
+			left: 0,
+			bottom: 0,
+			right: 0,
+			top: 0,
+		},
+		originOffset: {
+			x: 0,
+			y: 0,
+		}
+	};
+	private _workGlyphConfig: { chars: string | undefined; metadata: number } = { chars: undefined, metadata: 0 };
+
+	constructor(
+		private readonly _fontSize: number,
+		private readonly _fontFamily: string,
+	) {
+		super();
+
+		this._canvas = new OffscreenCanvas(this._fontSize * 3, this._fontSize * 3);
+		this._ctx = ensureNonNullable(this._canvas.getContext('2d', {
+			willReadFrequently: true
+		}));
+		this._ctx.textBaseline = 'top';
+		this._ctx.fillStyle = '#FFFFFF';
+	}
+
+	// TODO: Support drawing multiple fonts and sizes
+	/**
+	 * Rasterizes a glyph. Note that the returned object is reused across different glyphs and
+	 * therefore is only safe for synchronous access.
+	 */
+	public rasterizeGlyph(
+		chars: string,
+		metadata: number,
+		colorMap: string[],
+	): Readonly<IRasterizedGlyph> {
+		if (chars === '') {
+			return {
+				source: this._canvas,
+				boundingBox: { top: 0, left: 0, bottom: -1, right: -1 },
+				originOffset: { x: 0, y: 0 }
+			};
```

Comment:
This section here for `chars === ''` is only called a single time to fill in the "null" glyph. I think there's a issue to improve how the null glyph works. `_workGlyph` is shared across all other calls to avoid creating the objects at all. See the comment above:

```ts
	/**
	 * Rasterizes a glyph. Note that the returned object is reused across different glyphs and
	 * therefore is only safe for synchronous access.
	 */
```

I don't think your suggestion wrt class applies as it's always a single shape?

Reply:
Yes, if its a single instance it is a single shape.